add Namer for DNS SRV records

ccmtaylor opened the pull request:
This adds a plugin that allows linkerd to use SRV records for service discovery. A `linkerd.yaml` file using the plugin might contain something like this:

```yaml
routers:
- protocol: http
  dtab: |
    /dnssrv => /#/io.l5d.dnssrv;
    /svc/myservice =>
               /dnssrv/myservice.srv.example.org &
               /dnssrv/myservice2.srv.example.org;
    /svc/other => 
               /dnssrv/other.srv.example.org;

namers:
- kind: io.l5d.dnssrv
  experimental: true
```

Fixes #1610

hawkw commented:
Hi @ccmtaylor, this is great, thank you for your contribution!

I'll find some reviewers for this PR; in the mean time, would you mind making sure you've signed our [Contributor License Agreement](https://buoyant.io/cla/) so we can merge this?

hawkw commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,80 @@
+          NameTree.Leaf(Name.Bound(Var.value(Addr.Bound(srvRecords: _*)), id))
+        }
+      case code =>
+        log.trace("unexpected RCODE: %d for %s", code, address)
```
I think anything that makes `NameTree.Fail` probably ought to be logged at the `warning` level.

hawkw commented on namer/dnssrv/src/test/scala/io/buoyant/namer/dnssrv/DnsSrvNamerTest.scala:
```diff
@@ -0,0 +1,53 @@
+    assert(curator.dnsHosts === Some(Seq("localhost")))
+  }
+
+  test("can resolve some public SRV revord") {
```
I'd feel somewhat more comfortable if tests that create external network connections went in integration rather than unit tests.

hawkw commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,80 @@
+  extends Namer {
+
+  private val log = Logger.get("dnssrv")
+  private val cache = TrieMap.empty[Path, Var[State[NameTree[Name]]]]
```
Take it or leave it: consider [com.twitter.util.Memoize](https://twitter.github.io/util/docs/com/twitter/util/Memoize$.html#apply[A,B](f:A=>B):A=>B) as an alternative to `map.getOrElseUpdate`.

hawkw commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamerInitializer.scala:
```diff
@@ -0,0 +1,35 @@
+      case Some(hosts) => new DNS.ExtendedResolver(hosts.toArray)
+      case None => new DNS.ExtendedResolver()
+    }
+    resolver.setEDNS(0, 2048, 0, Collections.EMPTY_LIST)
```
nit: maybe put the level, payload size, etc, in a constant?

ccmtaylor commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -9,36 +9,28 @@ import com.twitter.util.Activity.State
     case _ => Activity.value(NameTree.Neg)
   }
 
-  private[dnssrv] def lookupSrv(address: String, id: Path, residual: Path): Try[NameTree[Name]] = Try {
```
I've seen other namers hand around a `residual` path, but I'm not sure how to use it, or why they would.

ccmtaylor commented:
@hawkw thaks for the quick review! I've addressed your comments; let me know if things look alright.

> would you mind making sure you've signed our Contributor License Agreement so we can merge this?

SoundCloud recently signed a CLA. I used my SC email in the commits, but if you prefer, I can close this PR and re-submit from a fork under github.com/soundcloud.

hawkw commented:
@ccmtaylor:
> SoundCloud recently signed a CLA. I used my SC email in the commits, but if you prefer, I can close this PR and re-submit from a fork under github.com/soundcloud.

Oh, okay; I wasn't aware of that and "please sign the CLA" is just in my boilerplate response for all first-time contributors. You're fine, then!

adleong commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,72 @@
+          // valid DNS entry, but no instances.
+          // for some reason, NameTree.Empty doesn't work right
+          log.trace("empty response for %s", address)
+          NameTree.Neg
```
It's debatable if you want NameTree.Neg here or a leaf with Addr.Bound containing an empty set.  The former will consider the name invalid and will fall back to any alternatives in the NameTree.  The later will be considered a valid name but any requests will fail because there are no addresses to send to.

ccmtaylor replied:
a major linkerd use-case for us is to load-balance across two or more deployments of a service (with different SRV records), i.e. the dtab looks like this:
```
    /svc/myservice =>
               /dnssrv/myservice1.srv.example.org &
               /dnssrv/myservice2.srv.example.org;
```
if `myservice1` returns `Addr.Bound(Set.empty)`, would linkerd still call `myservice2`?

adleong replied:
If you use an empty address set then half of the time you'll pick the first branch of the union and fail the request and half of the time you'll pick the second branch of the union and use myservice2

If you use NameTree.Neg then you'll skip to the myservice2 every time.

ccmtaylor replied:
sounds like `NameTree.Neg` is the right choice then :). In general, why would linkerd _not_ fallback/load-balance for empty sets? I see this has come up in #1612 and #1549, too.

adleong commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,72 @@
+          NameTree.Neg
+        } else {
+          log.trace("got %d results for %s", srvRecords.length, address)
+          NameTree.Leaf(Name.Bound(Var.value(Addr.Bound(srvRecords: _*)), id))
```
you want to set the residual as the `path` of the Name.Bound here

adleong commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,72 @@
+        }
+      case code =>
+        log.warning("unexpected RCODE: %s for %s", DNS.Rcode.string(code), address)
+        NameTree.Fail
```
I think this should be a `Throw(...)` instead of a `Return(NameTree.Fail)`

ccmtaylor replied:
ok. I'm a little unclear on the behaviour of the different `NameTree.*` variants. When would I use `NameTree.Fail`?

adleong replied:
It can be useful if you want to artificially halt evaluation of a name tree, but it's rarely used in practice.

adleong commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamerInitializer.scala:
```diff
@@ -0,0 +1,50 @@
+
+  @JsonIgnore
+  override def newNamer(params: Params): Namer = {
+    import org.xbill.DNS
```
move imports to top of file, please

adleong commented on namer/dnssrv/src/integration/scala/io/buoyant/namer/dnssrv/DnsSrvNamerIntegrationTest.scala:
```diff
@@ -0,0 +1,22 @@
+class DnsSrvNamerIntegrationTest extends FunSuite with Matchers {
+  test("can resolve some public SRV revord") {
+    val namer = new DnsSrvNamer(Path.empty, new DNS.ExtendedResolver, new NullTimer, Duration.Zero, new NullStatsReceiver)
+    val result = namer.lookupSrv("_http._tcp.mxtoolbox.com.", Path.read("/foo"))
```
I think this would be a better black-box test if you use namer.lookup instead.

ccmtaylor replied:
I agree, but I couldn't figure out how to do something like `Future.await()` for an `Activity`. The closest I could find is `Activity.sample`, but I'm not sure if that would race if the lookup is too slow. I'll push something, but please take a look if that's correct.

adleong replied:
if you import `io.buoyant.namer.RichActivity` then you can do `await(namer.lookup(...).toFuture)`

ccmtaylor commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,72 @@
+
+  private val log = Logger.get("dnssrv")
+  private val memoizedLookup: (Path) => Activity[NameTree[Name]] = Memoize {
+    case path@Path.Utf8(address, _) =>
```
ok. I actually wrote it that way first, but thought it looked a bit nicer like this because it's one fewer level of indentation (though I guess `Utf8.unapply()` might need to do more work.

ccmtaylor commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,72 @@
+    case _ => Activity.value(NameTree.Neg)
+  }
+
+  private[dnssrv] def lookupSrv(address: String, id: Path): Try[NameTree[Name]] = Try {
```
`resolver.send(query)` throws `IOException`. I'll scope the `Try{}` to that call.

ccmtaylor commented:
@hawkw @adleong: what kind of instrumentation would you expect on namers? I'm passing in a `StatsReceiver`, but not using it atm. I couldn't see anything consistent across the other namers, but I'm not sure if there exist some metrics across all kinds of namers (e.g. lookup latency, counts)?

ccmtaylor commented on namer/dnssrv/src/integration/scala/io/buoyant/namer/dnssrv/DnsSrvNamerIntegrationTest.scala:
```diff
@@ -0,0 +1,21 @@
+class DnsSrvNamerIntegrationTest extends FunSuite with Awaits with Matchers {
+  test("can resolve some public SRV revord") {
+    val namer = new DnsSrvNamer(Path.empty, new DNS.ExtendedResolver, new NullTimer, Duration.Zero, new NullStatsReceiver)
+    Activity.sample(namer.lookup(Path.read("/_http._tcp.mxtoolbox.com."))) match {
```
as mentioned in a previous comment, I'd appreciate a +1 that this shouldn't and doesn't just race on the `Activity` being in `Pending` state.

ccmtaylor commented on namer/dnssrv/src/test/scala/io/buoyant/namer/dnssrv/DnsSrvNamerTest.scala:
```diff
@@ -0,0 +1,38 @@
+  test("parse config") {
+    val yaml = s"""
+                  |kind: io.l5d.dnssrv
+                  |experimental: true
```
should I make this required? If so, where?

hawkw replied:
Set `experimentalRequired = true` in `DnsSrvNamerInitializer`. See (for example) `K8sNamerInitializer`:
https://github.com/linkerd/linkerd/blob/5d22b56489c4e646672e5d7fce6ef799d7574397/namer/k8s/src/main/scala/io/buoyant/namer/k8s/K8sExternalInitializer.scala#L34-L35

hawkw commented:
@ccmtaylor we don't currently collect any metrics from namers, although we probably _should_.

olix0r commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,76 @@
+    )
+    val query = DNS.Message.newQuery(question)
+    log.debug("looking up %s", address)
+    Try(resolver.send(query)) flatMap { message =>
```
I imagine that resolver.send is a potentially blocking call?  If so, then I think DnsSrvNamer should probably be constructed with a com.twitter.util.FuturePool so that these calls don't block the timer thread.

Furthermore, it may be a good idea to record a stat around the lookup times.

olix0r commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,76 @@
+              case Throw(e) => Activity.Failed(e)
+            }
+            state.update(next)
+          }
```
just to confirm i understand this... timer.schedule returns a closable and the timer is canceleld when the Var observation is released?

ccmtaylor replied:
yes, that is correct

ccmtaylor commented:
lifting this to a top-level comment, so it doesn't get lost in revisions:

> @olix0r: I imagine that resolver.send is a potentially blocking call? If so, then I think `DnsSrvNamer` should probably be constructed with a com.twitter.util.FuturePool so that these calls don't block the timer thread.

very good point, thanks @olix0r! I tried (and failed) to do this today. It's straight forward to make `lookupSrv` async with a `FuturePool` by doing

```scala
futurePool(resolver.send(query)) flatMap { message =>
  //...
  Future.value(NameTree.whatever)
  //...
  Future.exception(e)
}
```

but I can't figure out how to fit Futures into the Activity/Var/Timer APIs :(. Here's what I tried:

* I found `Activity.future(Future[T])`, but that seems to run a future once, and then stop.
* I tried to use `val (act, witness) = Activity[NameTree[Name]()` like this:
  ```scala
  // with lookupSrv modified to return Future[Nametree[Name]] as outlined above
  case id@Path.Utf8(address) =>
    val (act, witness) = Activity[NameTree[Name]()
    def loop(): Future[NameTree[Name]] = lookupSrv(address, prefix ++ id, path.drop(1)).respond { result =>
      witness.notify(result)
      timer.doLater(refreshInterval)(loop())
    }
    act
  ```
  but I'm left with no place to start and evaluate the loop -- calling `Await.result(loop())` would miss the point :)

I'm pretty fluent in using Futures, but the Activity and Var APIs are unfamiliar to me, so I'd appreciate any hints.

olix0r commented:
@ccmtaylor The Var/Activity apis are a bit nuanced and are difficult at times for all of us ;)  If you poke around some other namers you may find an idiom like the following:

```scala

implicit private[this] val timer = DefaultTimer.twitter

// ...
  Activity(Var.async[NameTree[Name]](NameTree.Pending) { state =>

     @volatile var stopped: Boolean = false

      def loop(): Future[Unit] = {
        if (stopped) Future.Unit
        else pool(resolve(name)).transform { result =>
            result match {
              Ok(addrs) =>
                state() = NameTree.Leaf(Addr.Bound(addrs))

              Throw(e) =>
                log.error(e, "resolution error: %s", name)
                // it may be appropriate to fail the resolution (i.e. Addr.Failed) if there was no good state before
            }

            // Wait a TTL before resolving again.
            Future.sleep(ttl).before(loop())
        }
      }

      val pending = loop()
      Closable.make { _ =>
        stopped = true

        // Cancel the timer notification or pending requests
        pending.raise(...)

        Future.Unit
      }
  })
...

... or something roughly like that.

ccmtaylor commented:
@olix0r thanks for the feedback, that worked great! I found `Future.whileDo`, which is a nice wrapper around the `loop`, and I used an `AtomicBoolean` as shown in the [twitter-util cookbook](https://twitter.github.io/util/guide/util-cookbook/futures.html#future-recursion).

olix0r commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,105 @@
+import com.twitter.util._
+import org.xbill.DNS
+
+class DnsSrvNamer(prefix: Path, resolver: DNS.Resolver, refreshInterval: Duration, stats: StatsReceiver, pool: FuturePool)(implicit val timer: Timer)
```
nit: prefer to break this very long line into a line per argument

olix0r commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,105 @@
+    val query = DNS.Message.newQuery(question)
+    log.debug("looking up %s", address)
+    pool {
+      val message = Stat.time(latency, TimeUnit.SECONDS)(resolver.send(query))
```
probably better to measure in milliseconds?

olix0r commented on namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala:
```diff
@@ -0,0 +1,105 @@
+  private val success = stats.counter("lookup_successes_total")
+  private val failure = stats.counter("lookup_failures_total")
+  private val zeroResults = stats.counter("lookup_zero_results_total")
+  private val latency = stats.stat("request_duration_seconds")
```
as mentioned below, this should probably be measured in millis (and have suffix `_ms` like other finagle metrics)

hawkw commented:
Hi @ccmtaylor, thanks for those last commits yesterday! I'm going to go ahead and merge this, and it'll show up in Linkerd 1.2.0 (which we're expecting to release by September 14th, if not before). Thanks again for your contribution!

ccmtaylor commented:
Thanks for your time and feedback, @hawkw, @adleong and @olix0r! I just noticed that I made an editing mistake while writing the docs. #1623  fixes it, sorry!

hawkw commented:
Thanks for the fix, @ccmtaylor!
